Add Queue Insights Documentation (#10400)
dotnet/arcade: Documentation/DevWorkflow/Queue-Insights-Documentation.md

# Queue Insights Documentation

Queue Insights is a GitHub check designed to show developers the current status of Helix, and information about what infrastructure a PR uses.

It's important to note that Queue Insights data is specific to the context of the PR. 

## Information Age

The information shown in Queue Insights is cached, updating daily, and consists of information over the past 30 days. Any queue or build machine used by your repo against a specific target branch will be shown for 30 days.

> This means that if your repo stops using a given Helix queue for instance, Queue Insights will show that queue for 30 days before it no longer appears in the check.

Reviewer: ```suggestion
This means that if your repo stops using a given Helix queue, Queue Insights will show that queue for 30 days before it no longer appears in the check.
```
The "for instance" is confusing to me. Feel free to edit if I misread.

## Pipeline Predictions

Here, we show the lower, expected, and higher estimated times for a pipeline the PR uses to complete, not including time spent in queue waiting for a build machine.

> You may interpret this data as this pipeline should take between the lower and upper bound, and commonly finishing around the estimated time.

Reviewer: ```suggestion
You should interpret this data as this pipeline should finish around the *expected* time but may finish as quick as the *lower* time or take as long as the *higher* time.
```
I think this is more clear? Open to debate though.

Author: I agree.

> This data is calculated from our build telemetry, and you may visit our [one-pager](../TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md) for how it works.

Reviewer: The one-pager links to a jupyter notebook hosted in ml.azure.com. What will happen to that after your account closed?

Reviewer: I kinda want that notebook file to be moved adjacent to the one-pager.

Author: I don't believe it should be deleted it's in the dnceng-internaltooling (?) subscription. So it shouldn't be deleted unless someone from the team deletes it.

 

### Multi-modal distributions
 
Some pipelines we cannot give an accurate prediction for. Typically, this is because the pipeline changes what stages run depending on the contents of the PR. Since we cannot account for this yet, you will experience an :x: symbol in the place of a prediction.

### Known Issues

> If a Critical Known Issue has been created, we will not show a prediction, and instead inform you that there is a critical known issue that could affect the estimated times.

Reviewer: ```suggestion
If a [Critical Known Issue](../Projects/Build%20Analysis/KnownIssues.md) has been created, we will not show a prediction, and instead inform you that there is a critical known issue that could affect the estimated times.
```
"Known Issue" is a special thing. Linking to a related doc should help with context.

Author: Good thinking! Thanks.

### Not enough data

If your pipeline is brand new or changed definition id's for instance, we won't have enough data to create predictions for. Currently, we require **30** runs of your pipeline before we can create an accurate enough model.

# Highest Work Item Wait Times

In this section, we show the top **5** (or less, depending on how many queues your PR uses) *work item wait times*. A work item is the smallest unit of work a queue can process in Helix. A Helix machine that processes work only runs one work item at a time.

*Work item wait time* is defined as the time a work item spends waiting to run. This is the amount of time the work item spends in queue waiting to run. A high work item wait time means that there are a lot of work items in that queue, and thus it might take longer to process work for that queue.

## Moving Average

> In this column, we show the moving average of the work item wait time, over the period of **24 hours**. This can be used to gauge if there is a sudden high influx of work items to a specific queue, which should be investigated with our *Grafana dashboard*

Reviewer: ```suggestion
In this column, we show the moving average of the work item wait time, over the period of **24 hours**. This can be used to gauge if there is a sudden high influx of work items to a specific queue, which should be investigated with our [Grafana Queue Monitor dashboard](https://dotnet-eng-grafana.westus2.cloudapp.azure.com/d/queues/queue-monitor).
```
Adding a link for context. I think this is the dashboard you're thinking of?

Reviewer: Eh, nevermind. The following section does the right thing.

## Grafana dashboard

Each queue mentioned in the Queue Insights check is hyperlinked, and will direct you to our Grafana dashboard **for that specific queue**. Here, you can dive deep into our infrastructure and view data like:

* Queue depth (the amount of work items waiting in that queue)
* Work item wait time
* Work items queued by repository
* Machine status (Busy processing work, provisioning, spinning down)

# Queue List

In this section, we show you a list of queues that this PR will use. **This is filtered by the queues the PR uses, not the overall repository.**. This is further separated by two factors, *on-premises* queues and *Azure queues*.

> ## :cloud: Helix Queues (Queues hosted in Azure)

Reviewer: The language in this and the next section is a bit conflicting. Helix Queues span both Azure and on-prem. So consider something like...

- :cloud: Azure-hosted queues
- :office: On-premises (build lab) queues

# Queue Insights Documentation

Queue Insights is a GitHub check designed to show developers the current status of Helix, and information about what infrastructure a PR uses.

It's important to note that Queue Insights data is specific to the context of the PR. 

## Information Age

The information shown in Queue Insights is cached, updating daily, and consists of information over the past 30 days. Any queue or build machine used by your repo against a specific target branch will be shown for 30 days.

This means that if your repo stops using a given Helix queue for instance, Queue Insights will show that queue for 30 days before it no longer appears in the check.

## Pipeline Predictions

Here, we show the lower, expected, and higher estimated times for a pipeline the PR uses to complete, not including time spent in queue waiting for a build machine.

You may interpret this data as this pipeline should take between the lower and upper bound, and commonly finishing around the estimated time.

> This data is calculated from our build telemetry, and you may visit our [one-pager]() for how it works.

Reviewer 2: I think you forgot about the link :(

 

### Multi-modal distributions
 
Some pipelines we cannot give an accurate prediction for. Typically, this is because the pipeline changes what stages run depending on the contents of the PR. Since we cannot account for this yet, you will experience an :x: symbol in the place of a prediction.

### Known Issues

> If there is a Critical Known Issue in our infrastructure (*e.g.* Helix, AzDo is down) we will not show a prediction, and instead inform you that there is a critical known issue that could affect the estimated ties.

Reviewer 3: Instead of mentioning the examples, we should just say that if a critical infrastructure issue has been filed. I think this opens us a bit to "Azdo is down, but you haven't created an issue yet, how come my PR still tried to show predictions?"
